File: tests/production/run_empirical.mjs
import ZAI from 'z-ai-web-dev-sdk';
import { writeFileSync, mkdirSync } from 'fs';

const zai = await ZAI.create();

async function chat(system, user, maxTokens = 300, temp = 0.1) {
  try {
    const completion = await zai.chat.completions.create({
      messages: [
        { role: 'system', content: system },
        { role: 'user', content: user }
      ],
      max_tokens: maxTokens,
      temperature: temp,
    });
    return completion.choices[0]?.message?.content?.trim() || '';
  } catch (e) {
    return '';
  }
}

const TICKETS = [
  { id: 'R-001', q: 'I want a refund for the headphones I bought 5 days ago. The left ear stopped working.', es: 'refund', ei: 'refund', em: 'neutral' },
  { id: 'R-002', q: 'Cancel my subscription immediately. I have been charged for 3 months and never used the service.', es: 'refund', ei: 'cancellation', em: 'angry' },
  { id: 'R-003', q: 'I returned my order 2 weeks ago but still have not received my money back. Order #ORD-88234.', es: 'refund', ei: 'refund', em: 'frustrated' },
  { id: 'R-004', q: 'This is the third time asking for a refund on the same item. Your system keeps rejecting it.', es: 'refund', ei: 'refund', em: 'angry' },
  { id: 'R-005', q: 'I bought a laptop 45 days ago. It is defective. Can I still get a refund?', es: 'refund', ei: 'refund', em: 'frustrated' },
  { id: 'R-006', q: 'My subscription renewed yesterday but I cancelled last week. I need the renewal charge reversed.', es: 'refund', ei: 'refund', em: 'frustrated' },
  { id: 'T-001', q: 'Your app keeps crashing when I try to upload files. Chrome version 125.', es: 'tech', ei: 'technical', em: 'neutral' },
  { id: 'T-002', q: 'The API is returning 503 errors intermittently. Our production integration is affected.', es: 'tech', ei: 'technical', em: 'urgent' },
  { id: 'T-003', q: 'I cannot log into my account. It says credentials invalid but I am using the right password.', es: 'tech', ei: 'technical', em: 'frustrated' },
  { id: 'T-004', q: 'The webhook integration stopped working after your last update. Events not being delivered.', es: 'tech', ei: 'technical', em: 'frustrated' },
  { id: 'T-005', q: 'My dashboard is loading extremely slow. 30 seconds for any page to load.', es: 'tech', ei: 'technical', em: 'frustrated' },
  { id: 'B-001', q: 'You charged me twice for the same order! Two charges of $149.99 on my card statement.', es: 'billing', ei: 'billing', em: 'angry' },
  { id: 'B-002', q: 'My invoice shows a different amount than what I was quoted. The tax calculation seems wrong.', es: 'billing', ei: 'billing', em: 'neutral' },
  { id: 'B-003', q: 'I upgraded from Starter to Pro last week but my invoice still shows the Starter price.', es: 'billing', ei: 'billing', em: 'frustrated' },
  { id: 'B-004', q: 'There is an unauthorized transaction of $3,450 on my account. Investigate immediately.', es: 'billing', ei: 'billing', em: 'angry' },
  { id: 'B-005', q: 'Can you explain the proration on my latest invoice? I do not understand the mid-cycle upgrade charge.', es: 'billing', ei: 'billing', em: 'neutral' },
  { id: 'G-001', q: 'What are your business hours? I need to know when I can reach a live agent.', es: 'general', ei: 'general', em: 'neutral' },
  { id: 'G-002', q: 'How do I change my email address on my account?', es: 'general', ei: 'account', em: 'neutral' },
  { id: 'G-003', q: 'I am very disappointed with the service. The agent was rude and unhelpful.', es: 'general', ei: 'complaint', em: 'angry' },
  { id: 'G-004', q: 'Do you offer an API for integrating with Salesforce?', es: 'general', ei: 'general', em: 'neutral' },
  { id: 'G-005', q: 'I am going to sue your company for selling my data without consent.', es: 'general', ei: 'escalation', em: 'angry' },
  { id: 'G-006', q: 'Can you tell me the status of my order #ORD-98234?', es: 'general', ei: 'order_status', em: 'neutral' },
  { id: 'G-007', q: 'I have been a loyal customer for 3 years and this is how you treat me? I want to speak to a manager.', es: 'general', ei: 'escalation', em: 'angry' },
  { id: 'G-008', q: 'What is the difference between your Starter and Pro plans?', es: 'general', ei: 'general', em: 'neutral' },
  { id: 'R-007', q: 'I accidentally purchased the Pro plan instead of Starter. Can you refund the difference?', es: 'refund', ei: 'refund', em: 'neutral' },
  { id: 'R-008', q: 'You people stole my money! I never signed up for this and you have been charging me for 6 months!', es: 'refund', ei: 'refund', em: 'angry' },
  { id: 'T-006', q: 'Getting SSL certificate errors when connecting to your API endpoint from our EU servers.', es: 'tech', ei: 'technical', em: 'neutral' },
  { id: 'T-007', q: 'Your mobile app will not open on my iPhone 15. Crashes immediately on launch.', es: 'tech', ei: 'technical', em: 'frustrated' },
  { id: 'B-006', q: 'My payment failed but my card is working fine everywhere else. What is wrong with your system?', es: 'billing', ei: 'billing', em: 'frustrated' },
  { id: 'B-007', q: 'I need a receipt for my annual subscription payment for tax purposes.', es: 'billing', ei: 'billing', em: 'neutral' },
];

const ROUTE_P = 'Classify into one word: refund/tech/billing/general. refund=money back/return/cancellation. tech=errors/bugs/API. billing=charges/invoices. general=everything else.';
const INTENT_P = 'Classify intent. One word: refund, cancellation, billing, technical, complaint, shipping, account, general, escalation, order_status.';
const RESP_P = {
  refund: 'You are a refund policy specialist. 30-day full refund policy, 31-60 day partial (50-75%), 60+ days only for defects. Subscription refunds are prorated from cancellation date. Be empathetic. Respond in 2-3 sentences with specific details about the refund.',
  tech: 'You are a technical support specialist. Start with simplest fix first. Give specific troubleshooting steps. If 3+ fixes fail, escalate to engineering. Respond in 2-3 sentences with actionable steps.',
  billing: 'You are a billing specialist. Verify charges against the subscription plan. Show exact amounts and line items. For disputes, investigate before crediting. Respond in 2-3 sentences with specific numbers.',
  general: 'You are a helpful customer support agent. Be friendly, clear, concise. For complaints, acknowledge frustration before solving. For legal threats, route to specialist team. Respond in 2-3 sentences.',
};
const EVAL_P = 'Judge if this customer support response actually resolves the problem. Be brutally honest from the customer perspective. Respond in JSON only: {"resolution_status": "fully_resolved" or "partially_resolved" or "not_resolved", "intent_match": true or false, "actionable": true or false, "reason": "brief explanation"}';

function parseEval(text) {
  try {
    let t = text.trim();
    if (t.includes('```json')) t = t.split('```json')[1].split('```')[0].trim();
    else if (t.includes('```')) t = t.split('```')[1].split('```')[0].trim();
    const d = JSON.parse(t);
    const status = ['fully_resolved','partially_resolved','not_resolved'].includes(d.resolution_status) ? d.resolution_status : 'not_resolved';
    return { status, intentMatch: !!d.intent_match, actionable: !!d.actionable, reason: d.reason || '' };
  } catch {
    let status = 'not_resolved';
    if (text.includes('fully_resolved')) status = 'fully_resolved';
    else if (text.includes('partially_resolved')) status = 'partially_resolved';
    return { status, intentMatch: false, actionable: false, reason: 'parse fallback' };
  }
}

const results = [];
const total = TICKETS.length;
console.log(`PARWA EMPIRICAL TEST: ${total} tickets via ZAI SDK`);
console.log('='.repeat(60));

for (let i = 0; i < total; i++) {
  const t = TICKETS[i];
  process.stdout.write(`[${i+1}/${total}] ${t.id} `);
  const start = Date.now();

  // Route
  const rSub = await chat(ROUTE_P, t.q, 20, 0.1);
  let predSub = 'general';
  for (const v of ['refund','tech','billing','general']) {
    if (rSub.toLowerCase().includes(v)) { predSub = v; break; }
  }

  // Intent
  const rInt = await chat(INTENT_P, t.q, 20, 0.1);
  let predInt = 'general';
  for (const v of ['refund','billing','technical','complaint','shipping','account','cancellation','general','escalation','order_status']) {
    if (rInt.toLowerCase().includes(v)) { predInt = v; break; }
  }

  // Response
  let sysP = RESP_P[predSub] || RESP_P.general;
  if (t.em === 'angry' || t.em === 'urgent') sysP += ' IMPORTANT: Customer is ANGRY. Show strong empathy first.';
  else if (t.em === 'frustrated') sysP += ' Customer is FRUSTRATED. Acknowledge their frustration.';
  const response = await chat(sysP, t.q, 400, 0.3);

  // Evaluate
  const evText = await chat(EVAL_P, `Expected intent: ${t.ei}\nCustomer: ${t.q}\nResponse: ${response}`, 200, 0.1);
  const ev = parseEval(evText);

  const lat = Date.now() - start;
  const subOk = predSub === t.es;
  const intOk = predInt === t.ei;
  const contained = !t.q.toLowerCase().includes('sue') && ev.status !== 'not_resolved';

  results.push({
    id: t.id, expectedSub: t.es, predSub, subOk,
    expectedIntent: t.ei, predInt, intOk,
    resolution: ev.status, intentMatch: ev.intentMatch,
    actionable: ev.actionable, reason: ev.reason,
    contained, response: response.slice(0, 200), latMs: lat,
  });

  const ri = { fully_resolved: 'F', partially_resolved: 'P', not_resolved: 'N' }[ev.status] || '?';
  console.log(`| Sub:${subOk?'S':'X'} Int:${intOk?'I':'X'} Res:${ri} | ${lat}ms | ${ev.reason?.slice(0,35)}`);

  // Small delay
  await new Promise(r => setTimeout(r, 500));
}

// METRICS
const n = results.length;
const containedN = results.filter(r => r.contained).length;
const subOkN = results.filter(r => r.subOk).length;
const intOkN = results.filter(r => r.intOk).length;
const fullyN = results.filter(r => r.resolution === 'fully_resolved').length;
const partialN = results.filter(r => r.resolution === 'partially_resolved').length;
const notResN = results.filter(r => r.resolution === 'not_resolved').length;
const trueResN = results.filter(r => r.intOk && r.resolution === 'fully_resolved').length;
const indResN = results.filter(r => r.contained && ['fully_resolved','partially_resolved'].includes(r.resolution)).length;
const intContN = results.filter(r => r.intOk && r.contained).length;
const evIntN = results.filter(r => r.intentMatch).length;
const evActionN = results.filter(r => r.actionable).length;

console.log();
console.log('='.repeat(60));
console.log('EMPIRICAL RESULTS - ZAI SDK - REAL LLM MEASUREMENTS');
console.log('='.repeat(60));
console.log(`Tickets: ${n}`);
console.log();
console.log(`Containment Rate:             ${(containedN/n*100).toFixed(1)}% (${containedN}/${n})`);
console.log(`Subgraph Routing Accuracy:    ${(subOkN/n*100).toFixed(1)}% (${subOkN}/${n})`);
console.log(`Intent Accuracy:              ${(intOkN/n*100).toFixed(1)}% (${intOkN}/${n})`);
console.log(`Evaluator Intent Match:       ${(evIntN/n*100).toFixed(1)}% (${evIntN}/${n})`);
console.log(`Actionable Response Rate:     ${(evActionN/n*100).toFixed(1)}% (${evActionN}/${n})`);
console.log(`Intent-Correct Containment:   ${(intContN/n*100).toFixed(1)}% (${intContN}/${n})`);
console.log(`Fully Resolved:               ${(fullyN/n*100).toFixed(1)}% (${fullyN}/${n})`);
console.log(`Partially Resolved:           ${(partialN/n*100).toFixed(1)}% (${partialN}/${n})`);
console.log(`Not Resolved:                 ${(notResN/n*100).toFixed(1)}% (${notResN}/${n})`);
console.log(`True Resolution Rate:         ${(trueResN/n*100).toFixed(1)}% (${trueResN}/${n})`);
console.log(`Industry-Comparable Rate:     ${(indResN/n*100).toFixed(1)}% (${indResN}/${n})`);

// By subgraph
const bySub = {};
results.forEach(r => { bySub[r.predSub] = bySub[r.predSub] || []; bySub[r.predSub].push(r); });
console.log('\nBY SUBGRAPH:');
for (const [sg, sgr] of Object.entries(bySub).sort()) {
  const sgn = sgr.length;
  const si = sgr.filter(r => r.intOk).length;
  const sf = sgr.filter(r => r.resolution === 'fully_resolved').length;
  const st = sgr.filter(r => r.intOk && r.resolution === 'fully_resolved').length;
  console.log(`  ${sg.padEnd(10)} ${sgn} tickets | Intent: ${(si/sgn*100).toFixed(0)}% | FullRes: ${(sf/sgn*100).toFixed(0)}% | TrueRes: ${(st/sgn*100).toFixed(0)}%`);
}

// Failed
const failed = results.filter(r => r.resolution === 'not_resolved' || !r.intOk);
if (failed.length) {
  console.log(`\nNEEDS IMPROVEMENT (${failed.length}):`);
  failed.forEach(r => {
    const im = r.intOk ? 'OK' : 'X';
    const sm = r.subOk ? 'OK' : 'X';
    console.log(`  ${r.id.padEnd(5)} Sub:${r.predSub.padEnd(8)}${sm} Int:${r.predInt.padEnd(12)}${im} Res:${r.resolution.padEnd(18)}`);
  });
}

// Levers
const intentFail = results.filter(r => !r.intOk).length;
const resFail = results.filter(r => r.resolution === 'not_resolved').length;
console.log(`\nIMPROVEMENT LEVERS:`);
console.log(`  Intent failures:  ${intentFail}/${n} (${(intentFail/n*100).toFixed(1)}%)`);
console.log(`  Resolution fails: ${resFail}/${n} (${(resFail/n*100).toFixed(1)}%)`);
console.log(`  BIGGEST LEVER: ${intentFail >= resFail ? 'intent' : 'response quality'}`);

// Industry comparison
console.log(`\nINDUSTRY COMPARISON:`);
console.log(`  Intercom:  50-70% claimed, ~40-55% real | PARWA: ${(indResN/n*100).toFixed(1)}%`);
console.log(`  Zendesk:   40-60% claimed, ~25-45% real | PARWA: ${(indResN/n*100).toFixed(1)}%`);
console.log(`  Sierra:    70-80% claimed, ~55-72% real | PARWA: ${(indResN/n*100).toFixed(1)}%`);
console.log(`  PARWA True Resolution: ${(trueResN/n*100).toFixed(1)}%`);

// Save
const output = {
  timestamp: new Date().toISOString(),
  method: 'empirical_zai_sdk',
  total_tickets: n,
  metrics: {
    containment_rate: +(containedN/n*100).toFixed(2),
    subgraph_routing_accuracy: +(subOkN/n*100).toFixed(2),
    intent_accuracy: +(intOkN/n*100).toFixed(2),
    evaluator_intent_match_rate: +(evIntN/n*100).toFixed(2),
    actionable_rate: +(evActionN/n*100).toFixed(2),
    intent_correct_containment_rate: +(intContN/n*100).toFixed(2),
    fully_resolved_rate: +(fullyN/n*100).toFixed(2),
    partially_resolved_rate: +(partialN/n*100).toFixed(2),
    not_resolved_rate: +(notResN/n*100).toFixed(2),
    true_resolution_rate: +(trueResN/n*100).toFixed(2),
    industry_comparable_rate: +(indResN/n*100).toFixed(2),
  },
  per_ticket_results: results,
};

mkdirSync('download', { recursive: true });
writeFileSync('download/empirical_resolution_rate_results.json', JSON.stringify(output, null, 2));
console.log('\nSaved to download/empirical_resolution_rate_results.json');

File: tests/production/run_empirical.py
"""
PARWA Empirical Resolution Rate - Inline Runner
Runs 20 tickets through real ZAI SDK LLM calls
"""
import asyncio
import json
import os
import re
import subprocess
import time
from datetime import datetime, timezone

_last_call_time = 0.0

async def zai_chat(system_prompt, user_message, max_retries=2):
    global _last_call_time
    for attempt in range(max_retries):
        try:
            now = time.monotonic()
            elapsed = now - _last_call_time
            if elapsed < 3.0:
                await asyncio.sleep(3.0 - elapsed)
            cmd = ["z-ai", "chat", "--prompt", user_message, "--system", system_prompt]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            _last_call_time = time.monotonic()
            if result.returncode == 0 and result.stdout.strip():
                try:
                    data = json.loads(result.stdout)
                    content = data.get("choices", [{}])[0].get("message", {}).get("content", "")
                    if content.strip():
                        return content.strip()
                except json.JSONDecodeError:
                    text = result.stdout.strip()
                    if text and len(text) > 5:
                        return text
            if attempt < max_retries - 1:
                await asyncio.sleep(3 * (attempt + 1))
            else:
                return ""
        except Exception as e:
            _last_call_time = time.monotonic()
            if attempt < max_retries - 1:
                await asyncio.sleep(3)
            else:
                return ""
    return ""

ROUTING_PROMPT = "Classify into one: refund/tech/billing/general. One word only.\n- refund: money back, return, reimbursement, cancellation\n- tech: errors, bugs, crashes, not working, API issues\n- billing: charges, payments, invoices, pricing\n- general: everything else"

INTENT_PROMPT = "Classify intent. One word only: refund, cancellation, billing, technical, complaint, shipping, account, general, escalation, order_status"

SUBGRAPH_PROMPTS = {
    "refund": "You are a refund policy specialist. 30-day full refund, 31-60 partial, 60+ defects only. Be empathetic. Respond professionally in 2-4 sentences.",
    "tech": "You are a tech support specialist. Start with simplest fix. If 3+ fixes fail, escalate. Respond professionally in 2-4 sentences.",
    "billing": "You are a billing specialist. Verify charges against plan. Show exact line items. Respond professionally in 2-4 sentences.",
    "general": "You are a helpful support agent. Be friendly, clear, concise. Respond professionally in 2-4 sentences.",
}

EVALUATOR_PROMPT = "Judge if this response resolves the customer problem. Be brutally honest. Respond in JSON format with keys: resolution_status (fully_resolved/partially_resolved/not_resolved), intent_match (true/false), reason (brief string)."


async def route_ticket(query):
    r = await zai_chat(ROUTING_PROMPT, query)
    r = r.strip().lower()
    for v in ("refund", "tech", "billing", "general"):
        if v in r:
            return v
    return "general"


async def classify_intent(query):
    r = await zai_chat(INTENT_PROMPT, query)
    r = r.strip().lower()
    for v in ("refund", "billing", "technical", "complaint", "shipping", "account", "cancellation", "general", "escalation", "order_status"):
        if v in r:
            return v
    return "general"


async def generate_response(query, subgraph, emotion):
    system = SUBGRAPH_PROMPTS.get(subgraph, SUBGRAPH_PROMPTS["general"])
    if emotion in ("angry", "urgent"):
        system += " Customer is ANGRY. Show strong empathy first."
    elif emotion == "frustrated":
        system += " Customer is FRUSTRATED. Acknowledge frustration."
    return await zai_chat(system, query)


async def evaluate(query, response, expected_intent):
    r = await zai_chat(EVALUATOR_PROMPT, "Expected intent: " + expected_intent + "\nCustomer: " + query + "\nResponse: " + response)
    try:
        text = r.strip()
        # Remove code fences
        if "```json" in text:
            text = text.split("```json")[1].split("```")[0].strip()
        elif "```" in text:
            text = text.split("```")[1].split("```")[0].strip()
        parsed = json.loads(text)
        status = parsed.get("resolution_status", "not_resolved")
        if status not in ("fully_resolved", "partially_resolved", "not_resolved"):
            status = "not_resolved"
        return {
            "resolution_status": status,
            "intent_match": bool(parsed.get("intent_match", False)),
            "reason": parsed.get("reason", ""),
        }
    except (json.JSONDecodeError, ValueError):
        status = "not_resolved"
        if "fully_resolved" in r:
            status = "fully_resolved"
        elif "partially_resolved" in r:
            status = "partially_resolved"
        return {"resolution_status": status, "intent_match": False, "reason": "parse fallback"}


TICKETS = [
    {"id": "R-001", "query": "I want a refund for the headphones I bought 5 days ago. The left ear stopped working.", "category": "refund", "expected_subgraph": "refund", "expected_intent": "refund", "emotion": "neutral"},
    {"id": "R-002", "query": "Cancel my subscription immediately. I have been charged for 3 months and never used the service.", "category": "cancellation", "expected_subgraph": "refund", "expected_intent": "cancellation", "emotion": "angry"},
    {"id": "R-003", "query": "I returned my order 2 weeks ago but still have not received my money back. Order #ORD-88234.", "category": "refund", "expected_subgraph": "refund", "expected_intent": "refund", "emotion": "frustrated"},
    {"id": "R-004", "query": "This is the third time asking for a refund on the same item. Your system keeps rejecting it.", "category": "refund", "expected_subgraph": "refund", "expected_intent": "refund", "emotion": "angry"},
    {"id": "R-005", "query": "I bought a laptop 45 days ago. It is defective. Can I still get a refund?", "category": "refund", "expected_subgraph": "refund", "expected_intent": "refund", "emotion": "frustrated"},
    {"id": "T-001", "query": "Your app keeps crashing when I try to upload files. I am on Chrome version 125.", "category": "technical", "expected_subgraph": "tech", "expected_intent": "technical", "emotion": "neutral"},
    {"id": "T-002", "query": "The API is returning 503 errors intermittently. Our production integration is affected.", "category": "technical", "expected_subgraph": "tech", "expected_intent": "technical", "emotion": "urgent"},
    {"id": "T-003", "query": "I cannot log into my account. It says my credentials are invalid but I am using the right password.", "category": "technical", "expected_subgraph": "tech", "expected_intent": "technical", "emotion": "frustrated"},
    {"id": "T-004", "query": "The webhook integration stopped working after your last update. Events are not being delivered.", "category": "technical", "expected_subgraph": "tech", "expected_intent": "technical", "emotion": "frustrated"},
    {"id": "B-001", "query": "You charged me twice for the same order. I can see two charges of 149.99 on my card statement.", "category": "billing", "expected_subgraph": "billing", "expected_intent": "billing", "emotion": "angry"},
    {"id": "B-002", "query": "My invoice shows a different amount than what I was quoted. The tax calculation seems wrong.", "category": "billing", "expected_subgraph": "billing", "expected_intent": "billing", "emotion": "neutral"},
    {"id": "B-003", "query": "I upgraded from Starter to Pro last week but my invoice still shows the Starter price.", "category": "billing", "expected_subgraph": "billing", "expected_intent": "billing", "emotion": "frustrated"},
    {"id": "B-004", "query": "There is an unauthorized transaction of 3450 on my account. I need this investigated immediately.", "category": "billing", "expected_subgraph": "billing", "expected_intent": "billing", "emotion": "angry"},
    {"id": "G-001", "query": "What are your business hours? I need to know when I can reach a live agent.", "category": "general", "expected_subgraph": "general", "expected_intent": "general", "emotion": "neutral"},
    {"id": "G-002", "query": "How do I change my email address on my account?", "category": "account", "expected_subgraph": "general", "expected_intent": "account", "emotion": "neutral"},
    {"id": "G-003", "query": "I am very disappointed with the service I received. The agent was rude and unhelpful.", "category": "complaint", "expected_subgraph": "general", "expected_intent": "complaint", "emotion": "angry"},
    {"id": "G-004", "query": "Do you offer an API for integrating with Salesforce?", "category": "general", "expected_subgraph": "general", "expected_intent": "general", "emotion": "neutral"},
    {"id": "G-005", "query": "I am going to sue your company for selling my data without consent.", "category": "escalation", "expected_subgraph": "general", "expected_intent": "escalation", "emotion": "angry"},
    {"id": "G-006", "query": "Can you tell me the status of my order #ORD-98234?", "category": "order_status", "expected_subgraph": "general", "expected_intent": "order_status", "emotion": "neutral"},
    {"id": "G-007", "query": "I have been a loyal customer for 3 years and this is how you treat me? I want to speak to a manager.", "category": "escalation", "expected_subgraph": "general", "expected_intent": "escalation", "emotion": "angry"},
]


async def run_all():
    results = []
    total = len(TICKETS)
    print("PARWA EMPIRICAL TEST: %d tickets, real ZAI SDK LLM calls" % total)
    print("=" * 70)

    for i, t in enumerate(TICKETS):
        tid = t["id"]
        print("[%d/%d] %s " % (i+1, total, tid), end="", flush=True)
        start = time.monotonic()

        pred_sub = await route_ticket(t["query"])
        await asyncio.sleep(1)
        pred_int = await classify_intent(t["query"])
        await asyncio.sleep(1)
        response = await generate_response(t["query"], pred_sub, t["emotion"])
        await asyncio.sleep(1)
        ev = await evaluate(t["query"], response, t["expected_intent"])

        latency = round((time.monotonic() - start) * 1000)

        sub_ok = pred_sub == t["expected_subgraph"]
        int_ok = pred_int == t["expected_intent"]
        # Legal threats should escalate
        contained = "sue" not in t["query"].lower()

        r = {
            "ticket_id": tid,
            "category": t["category"],
            "expected_subgraph": t["expected_subgraph"],
            "predicted_subgraph": pred_sub,
            "subgraph_correct": sub_ok,
            "expected_intent": t["expected_intent"],
            "predicted_intent": pred_int,
            "intent_correct": int_ok,
            "resolution_status": ev["resolution_status"],
            "evaluator_intent_match": ev["intent_match"],
            "evaluator_reason": ev["reason"],
            "contained": contained,
            "response": response[:200],
            "latency_ms": latency,
        }
        results.append(r)

        ri = {"fully_resolved": "F", "partially_resolved": "P", "not_resolved": "N"}.get(ev["resolution_status"], "?")
        sub_mark = "S" if sub_ok else "X"
        int_mark = "I" if int_ok else "X"
        print("| Sub:%s Int:%s Res:%s | %dms" % (sub_mark, int_mark, ri, latency))

        await asyncio.sleep(1)

    # Calculate metrics
    n = len(results)
    contained_n = len([r for r in results if r["contained"]])
    sub_ok_n = len([r for r in results if r["subgraph_correct"]])
    int_ok_n = len([r for r in results if r["intent_correct"]])
    fully_n = len([r for r in results if r["resolution_status"] == "fully_resolved"])
    partial_n = len([r for r in results if r["resolution_status"] == "partially_resolved"])
    not_res_n = len([r for r in results if r["resolution_status"] == "not_resolved"])
    true_res_n = len([r for r in results if r["intent_correct"] and r["resolution_status"] == "fully_resolved"])
    ind_res_n = len([r for r in results if r["contained"] and r["resolution_status"] in ("fully_resolved", "partially_resolved")])
    int_contained_n = len([r for r in results if r["intent_correct"] and r["contained"]])
    ev_int_n = len([r for r in results if r["evaluator_intent_match"]])

    # Per-subgraph breakdown
    by_sub = {}
    for r in results:
        by_sub.setdefault(r["predicted_subgraph"], []).append(r)

    print()
    print("=" * 70)
    print("EMPIRICAL RESULTS - REAL LLM MEASUREMENTS")
    print("=" * 70)
    print("Tickets: %d | LLM calls: ~%d | Model: GLM-4-Plus" % (n, n*5))
    print()
    print("INDUSTRY-STANDARD METRICS:")
    print("  Method 1: Containment Rate:              %.1f%%  (%d/%d)" % (contained_n/n*100, contained_n, n))
    print("  Method 2: Intent-Correct Containment:    %.1f%%  (%d/%d)" % (int_contained_n/n*100, int_contained_n, n))
    print("  Method 3: True Resolution Rate:          %.1f%%  (%d/%d)" % (true_res_n/n*100, true_res_n, n))
    print("  Method 4: Industry-Comparable Rate:      %.1f%%  (%d/%d)" % (ind_res_n/n*100, ind_res_n, n))
    print()
    print("BREAKDOWN:")
    print("  Subgraph Routing Accuracy:  %.1f%%" % (sub_ok_n/n*100))
    print("  Intent Accuracy:            %.1f%%" % (int_ok_n/n*100))
    print("  Evaluator Intent Match:     %.1f%%" % (ev_int_n/n*100))
    print("  Fully Resolved:             %.1f%%" % (fully_n/n*100))
    print("  Partially Resolved:         %.1f%%" % (partial_n/n*100))
    print("  Not Resolved:               %.1f%%" % (not_res_n/n*100))
    print()
    print("BY SUBGRAPH:")
    for sg, sg_results in sorted(by_sub.items()):
        sg_n = len(sg_results)
        sg_int = len([r for r in sg_results if r["intent_correct"]])
        sg_full = len([r for r in sg_results if r["resolution_status"] == "fully_resolved"])
        sg_true = len([r for r in sg_results if r["intent_correct"] and r["resolution_status"] == "fully_resolved"])
        sg_cont = len([r for r in sg_results if r["contained"]])
        print("  %-10s %d tickets | Intent: %.0f%% | Contain: %.0f%% | TrueRes: %.0f%%" % (
            sg, sg_n, sg_int/sg_n*100, sg_cont/sg_n*100, sg_true/sg_n*100))

    # Failed tickets
    failed = [r for r in results if r["resolution_status"] == "not_resolved" or not r["intent_correct"]]
    if failed:
        print()
        print("TICKETS NEEDING IMPROVEMENT (%d):" % len(failed))
        for r in failed:
            int_mark = "X" if not r["intent_correct"] else "OK"
            sub_mark = "X" if not r["subgraph_correct"] else "OK"
            print("  %-5s [%-12s] Sub:%-8s %s Int:%-12s %s | Res:%-18s | %s" % (
                r["ticket_id"], r["category"],
                r["predicted_subgraph"], sub_mark,
                r["predicted_intent"], int_mark,
                r["resolution_status"],
                r["evaluator_reason"][:50]))

    # Improvement levers
    intent_fail = [r for r in results if not r["intent_correct"]]
    res_fail = [r for r in results if r["resolution_status"] == "not_resolved"]
    print()
    print("IMPROVEMENT LEVERS:")
    print("  Intent failures:    %d/%d (%.1f%%)" % (len(intent_fail), n, len(intent_fail)/n*100))
    print("  Resolution failures: %d/%d (%.1f%%)" % (len(res_fail), n, len(res_fail)/n*100))
    biggest = "intent" if len(intent_fail) >= len(res_fail) else "resolution quality"
    print("  BIGGEST LEVER: %s" % biggest)

    # Industry comparison
    print()
    print("INDUSTRY COMPARISON:")
    print("  %-20s %10s %10s %10s" % ("Company", "Reported", "Real Est.", "PARWA"))
    print("  %-20s %10s %10s %10s" % ("Intercom Fin", "50-70%", "35-55%", "%.1f%%" % (ind_res_n/n*100)))
    print("  %-20s %10s %10s %10s" % ("Zendesk AI", "40-60%", "25-45%", "%.1f%%" % (ind_res_n/n*100)))
    print("  %-20s %10s %10s %10s" % ("Sierra AI", "70-80%", "55-72%", "%.1f%%" % (ind_res_n/n*100)))
    print("  %-20s %10s %10s %10s" % ("PARWA (true)", "--", "--", "%.1f%%" % (true_res_n/n*100)))

    # Save
    output = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "method": "empirical_real_llm_zai_sdk_glm4_plus",
        "total_tickets": n,
        "total_llm_calls_approx": n * 5,
        "metrics": {
            "containment_rate": round(contained_n/n*100, 2),
            "subgraph_routing_accuracy": round(sub_ok_n/n*100, 2),
            "intent_accuracy": round(int_ok_n/n*100, 2),
            "evaluator_intent_match_rate": round(ev_int_n/n*100, 2),
            "intent_correct_containment_rate": round(int_contained_n/n*100, 2),
            "fully_resolved_rate": round(fully_n/n*100, 2),
            "partially_resolved_rate": round(partial_n/n*100, 2),
            "not_resolved_rate": round(not_res_n/n*100, 2),
            "true_resolution_rate": round(true_res_n/n*100, 2),
            "industry_comparable_rate": round(ind_res_n/n*100, 2),
        },
        "per_ticket_results": results,
        "improvement_levers": {
            "intent_failures": len(intent_fail),
            "resolution_failures": len(res_fail),
            "biggest_lever": biggest,
        },
    }

    os.makedirs("download", exist_ok=True)
    with open("download/empirical_resolution_rate_results.json", "w") as f:
        json.dump(output, f, indent=2, default=str)
    print("\nSaved to download/empirical_resolution_rate_results.json")


if __name__ == "__main__":
    asyncio.run(run_all())
